SourceX/controls/plrctrls.cpp

```diff
@@ -0,0 +1,693 @@
+			spbslot = r;
+			//sprintf(tempstr, "IN HOT SPELL CELL NUM:%i", r);
+			//NetSendCmdString(1 << myplr, tempstr);
+		}
+	}
+
+	if (key == DVL_VK_UP) {
+		if (speedspellscoords[spbslot].y == 307 && hsr[1] > 0) { // we're in row 1, check if row 2 has spells
+			if (HSExists(MouseX, 251)) {
+				x = MouseX;
+				y = 251;
+			}
+		} else if (speedspellscoords[spbslot].y == 251 && hsr[2] > 0) { // we're in row 2, check if row 3 has spells
+			if (HSExists(MouseX, 195)) {
+				x = MouseX;
+				y = 195;
+			}
+		}
+	} else if (key == DVL_VK_DOWN) {
+		if (speedspellscoords[spbslot].y == 251) { // we're in row 2
+			if (HSExists(MouseX, 307)) {
+				x = MouseX;
+				y = 307;
+			}
+		} else if (speedspellscoords[spbslot].y == 195) { // we're in row 3
+			if (HSExists(MouseX, 251)) {
+				x = MouseX;
+				y = 251;
+			}
+		}
+	} else if (key == DVL_VK_LEFT) {
+		if (spbslot >= speedspellcount - 1)
+			return;
+		spbslot++;
+		x = speedspellscoords[spbslot].x;
+		y = speedspellscoords[spbslot].y;
+	} else if (key == DVL_VK_RIGHT) {
+		if (spbslot <= 0)
+			return;
+		spbslot--;
+		x = speedspellscoords[spbslot].x;
+		y = speedspellscoords[spbslot].y;
+	}
+
+	if (x > 0 && y > 0) {
+		SetCursorPos(x, y);
+		MouseX = x;
+		MouseY = y;
+	}
+}
+
+void walkInDir(MoveDirection dir)
+{
+	if (dir.x == MoveDirectionX::NONE && dir.y == MoveDirectionY::NONE)
+		return;
+	DWORD ticks = GetTickCount();
+	if (ticks - invmove < 370) {
+		return;
+	}
+	invmove = ticks;
+	ClrPlrPath(myplr);                   // clear nodes
+	plr[myplr].destAction = ACTION_NONE; // stop attacking, etc.
+	HideCursor();
+	static const _walk_path kMoveToWalkDir[3][3] = {
+		// NONE      UP      DOWN
+		{ WALK_NONE, WALK_N, WALK_S }, // NONE
+		{ WALK_W, WALK_NW, WALK_SW },  // LEFT
+		{ WALK_E, WALK_NE, WALK_SE },  // RIGHT
+	};
+	plr[myplr].walkpath[0] = kMoveToWalkDir[static_cast<std::size_t>(dir.x)][static_cast<std::size_t>(dir.y)];
+}
+
+void menuMoveX(MoveDirectionX dir)
+{
+	if (dir == MoveDirectionX::NONE)
+		return;
+	invMove(dir == MoveDirectionX::LEFT ? WALK_W : WALK_E);
+	hotSpellMove(dir == MoveDirectionX::LEFT ? DVL_VK_LEFT : DVL_VK_RIGHT);
+}
+
+void menuMoveY(MoveDirectionY dir)
+{
+	if (dir == MoveDirectionY::NONE)
+		return;
+	invMove(dir == MoveDirectionY::UP ? WALK_N : WALK_S);
+	const auto key = dir == MoveDirectionY::UP ? DVL_VK_UP : DVL_VK_DOWN;
+	hotSpellMove(key);
+	attrIncBtnSnap(key);
+}
+
+void movement()
+{
+	if (InGameMenu())
+		return;
+
+	MoveDirection move_dir = GetMoveDirection();
+	if (InControlledMenu()) {
+		menuMoveX(move_dir.x);
+		menuMoveY(move_dir.y);
+	} else {
+		walkInDir(move_dir);
+	}
+
+	if (GetAsyncKeyState(DVL_MK_LBUTTON) & 0x8000) {
+		if (newCurHidden) { // show cursor first, before clicking
+			SetCursor_(CURSOR_HAND);
+			newCurHidden = false;
+		} else if (spselflag) {
+			SetSpell();
+		} else {
+			LeftMouseCmd(false);
+		}
+	}
+}
+
+// Move map or mouse no more than 60 times per second.

```

Comment:
Maybe better to move x px per sec instead of locking it to 60fps, the game runs at vsync, so many monitors will allow for more then 60fps.